### DISCUSSION TITLE
[CSV] Add `rename` option

### DISCUSSION THREAD

**Discussion by Tishj** [duckdb/duckdb#12536] [Ideas]
In the form of `MAP(VARCHAR, VARCHAR)` where the key is the existing name in the file, and the value is the new name of the column

Eg:
`read_csv('my_file.csv', rename = {'original': 'renamed'})`

This fills in the gap where `names` or `columns` can not be used because not all columns are known up front

**Comment by tonne1** [comment]
Good idea to make these things easier. Although I wonder if mappings like that couldn't be unified, so that the same mechanism could be applied to csv, json, parquet or regular table data sets. There is a number of different approaches to deal with columns. From https://duckdb.org/docs/sql/expressions/star:

"The matches of capture groups can be used to rename columns selected by a regular expression:"
```sql
SELECT COLUMNS('(\w{2}).*') AS '\1' FROM numbers;
```
This isn't really rename, but it replaces the original with the contents of a capture group. But at least it transforms the column names. And there is
```sql
SELECT COLUMNS(c -> c LIKE '%num%') FROM numbers;
```
but in this context, lambdas are only interpreted as a filter and an attempt to do e.g.
```sql
select columns(c -> c.replace('original', 'renamed')) from read_csv('my_file.csv');
```
Conversion Error: Could not convert string 'renamed' to BOOL.
Another idea, using the replace option of columns(*):
```sql
select columns(* replace(original, renamed)) from read_csv('my_file.csv');
```
Binder Error: Column "renamed" in REPLACE list not found in FROM clause. Replace operates on the expression, not on the name.
Now if we had RENAME as a third option with columns, which operates on the name/column alias instead:
```sql
select columns(* rename(original, renamed)) from read_csv('my_file.csv');
```
That would match the current columns handling. Of course things can get tricky if these options are combined. 
Actually I wonder what kind of data structure "columns" effectively is.

**Reply by Tishj** [reply]
Thanks, the problem with COLUMNS is that it acts too late for the purpose I'm thinking of.
But maybe that purpose is too niche to warrant any development time and added confusion.

The thing that inspired this post was issues like `filename`, where we can add an artificial column that might conflict with existing columns.
But this is already being solved the other way by allowing you to rename the artificial column we add (in the case of `filename` starting with the next release)
